platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -278,15 +279,26 @@ private static DefaultJvmSoftwareComponent createJavaComponent(Project project,
 
         // Supply the sourceSet to the feature
         JvmFeatureInternal feature = new DefaultJvmFeature(
-            JvmConstants.JAVA_FEATURE_NAME, sourceSet, Collections.emptyList(),
+            JvmConstants.JAVA_MAIN_FEATURE_NAME, sourceSet, Collections.emptyList(),
             (ProjectInternal) project, false, false);
+        // Create a source directories variant for the feature
+        feature.withSourceElements();
+
+        // Build the main jar when running `assemble`.
+        DefaultArtifactPublicationSet publicationSet = project.getExtensions().getByType(DefaultArtifactPublicationSet.class);
+        publicationSet.addCandidate(feature.getRuntimeElementsConfiguration().getArtifacts().iterator().next());
 
         // And supply main feature to the component
         DefaultJvmSoftwareComponent component = project.getObjects().newInstance(
             DefaultJvmSoftwareComponent.class,
-            JvmConstants.JAVA_COMPONENT_NAME,
-            project,
-            feature
+            JvmConstants.JAVA_MAIN_COMPONENT_NAME,
+            feature,
+            project.getObjects(),
```

Comment:
The object factory should auto-inject all these parameters. We should only need to provide the parameters that that are not injectable from the services

Reply:
All the `project.` parameters should be able to be omitted. Applies to the tests as well. 

No need to hold up the merge on this, but we should clean this up in a follow-up

Reply:
Merge conflict kicked this back, so I'll update this as I rebase.